merge.py

```diff
@@ -1,98 +1,196 @@
+             cwd=tmp_dir)
+        call(['git', 'remote', 'rename', 'origin', SOURCE_REMOTE], cwd=tmp_dir)
+        call('git config --local --add remote.' + SOURCE_REMOTE +
+             '.fetch "+refs/pull/*/head:refs/remotes/{}/pr/*"'.format(
+                 SOURCE_REMOTE), shell=True,
+             cwd=tmp_dir)
+        call(['git', 'fetch', '--all'], cwd=tmp_dir)
+        _print_flush('Initial work complete.')
+
+        # Clean up fetch
+        pr_name = 'finish-pr-{}'.format(pr)
+        ic_args = ['git', 'checkout', '-b', pr_name, 'github/pr/{}'.format(pr)]
+        initial_checkout = call(ic_args, cwd=tmp_dir)
+        if initial_checkout != 0:
+            pr.post_error("Couldn't checkout code. Please try again.")
+            return False
+        _print_flush('Checked out.')
+
+        # Rebase PR onto main.
+        branch = '{}/{}'.format(TARGET_REMOTE, TARGET_BRANCH)
+        rebase_args = ['git', 'rebase', branch]
+        rebase_success = call(rebase_args, cwd=tmp_dir)
+        if rebase_success != 0:
+            _print_flush(rebase_success)
+            pr.post_error('Rebase was not successful. Please rebase against '
+                          'main and try again.')
+            return False
+        _print_flush('Rebased')
+
+        # Check out target branch to here
+        checkout_success = call(['git', 'checkout', branch], cwd=tmp_dir)
+        if checkout_success != 0:
+            pr.post_error(
+                'Error checking out target branch: master. Please try again.')
+            return False
+        _print_flush('Checked out Apache master.')
+
+        # Merge
+        merge_msg = 'This closes #{}'.format(pr)
+        merge_args = ['git', 'merge', '--no-ff', '-m', merge_msg, pr_name]
+        merge_success = call(merge_args, cwd=tmp_dir)
+        if merge_success != 0:
+            pr.post_error(
+                'Merge was not successful against target branch: master. '
+                'Please try again.')
+            return False
+        _print_flush('Merged successfully.')
+
+        _print_flush('Running mvn clean verify.')
+        # mvn clean verify
+        mvn_success = call(['mvn', 'clean', 'verify'], cwd=tmp_dir)
+        if mvn_success != 0:
+            pr.post_error(
+                'verify against HEAD + PR#{} failed. Not merging.'.format(pr))
+            return False
+
+        # git push
+        push_args = ['git', 'push', 'apache', 'HEAD:master']
+        push_success = call(push_args, cwd=tmp_dir)
+        if push_success != 0:
+            pr.post_error('Git push failed. Please try again.')
+            return False
+        return True
+
+
+def _set_up(tmp_dir):
+    """set_up creates a temp directory for use.
+
+    Args:
+        tmp_dir: Directory to create.
+    Returns:
+        True if directory was created successfully.
+        False otherwise.
+    """
+    if call(['mkdir', tmp_dir]) != 0:
+        return False
+    return True
+
+
+def _clean_up(tmp_dir):
```

Comment:
Trying to decide if it goes inside merge_github_pr or not. If we fail to set up or clean up I'm not sure what the failure mode should be.